src/matchbox/common/db.py

```diff
@@ -136,6 +236,94 @@ def __hash__(self) -> int:
+            2. Columns specified in TOML - uses specified columns with optional '*'
+            3. Indices from database - uses existing column hash information
+        """
+        # Initialise warehouse and get table metadata
+        warehouse = (
+            data["database"]
+            if isinstance(data["database"], SourceWarehouse)
+            else SourceWarehouse(**data["database"])
+        )
+
+        metadata = MetaData(schema=data["db_schema"])
+        table = Table(data["db_table"], metadata, autoload_with=warehouse.engine)
+
+        # Get all columns except primary key
+        remote_columns = [
+            SourceColumn(literal=col.name, type=str(col.type), indexed=False)
+            for col in table.columns
+            if col.name not in data["db_pk"]
+        ]
+
+        index_data = data.get("index")
+
+        # Case 1: No columns specified - index everything
+        if not index_data:
+            data["db_columns"] = [
+                SourceColumn(literal=col.literal.name, type=col.type, indexed=True)
+                for col in remote_columns
+            ]
+            return data
+
+        # Case 2: Columns from database
+        if isinstance(index_data, dict):
+            source_index = SourceIndex(**index_data)
+            data["db_columns"] = [
+                SourceColumn(
+                    literal=col.literal.name,
+                    type=col.type,
+                    indexed=col in source_index.literal + source_index.alias,
+                )
+                for col in remote_columns
+            ]
+            return data
+
+        # Case 3: Columns from TOML
+        local_columns = []
+        star_index = None
+
+        # Process TOML column specifications
+        for i, column in enumerate(index_data):
+            if column["literal"] == "*":
+                star_index = i
+                continue
+            local_columns.append(SourceColumn(**column, indexed=True))
+
+        # Match remote columns with local specifications
+        indexed_columns = []
+        non_indexed_columns = []
+
+        for remote_col in remote_columns:
+            matched = False
+            for local_col in local_columns:
+                if remote_col == local_col:
+                    if local_col.type is None:
+                        local_col.type = remote_col.type
+                    indexed_columns.append(local_col)
+                    matched = True
+                    break
+            if not matched:
+                non_indexed_columns.append(remote_col)
+
+        # Handle wildcard insertion
+        if star_index is not None:
+            for col in non_indexed_columns:
+                col.indexed = True
+            indexed_columns[star_index:star_index] = non_indexed_columns
+            data["db_columns"] = indexed_columns
+        else:
+            data["db_columns"] = indexed_columns + non_indexed_columns
```

Comment:
My comment above applies here as well: is it worth being more conservative? How much are we gaining by having all these different options to index?

Reply:
What could happen is that the TOML could stay the same, but the table could change, and this would make the indexing of the hash of the table change. Stability of the hash as defined by the TOML feels like a feature, which the "star" breaks